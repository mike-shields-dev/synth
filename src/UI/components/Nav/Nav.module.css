.Nav {
    container-type: inline-size;
    grid-area: header;
    background: rgba(0, 0, 0, 0.125);
    display: flex;
    width: 100%;
    height: 100%;
    padding: 0.25rem;
    justify-content: center;
}

.Nav a {
    font-size: 0.75em;
    display: grid;
    place-items: center;
    border: none;
    border-radius: 0.25em;
    background: rgba(255, 255, 255, 0.125);
    padding: 0.125em 0.25em;
    color: #fff;
    text-decoration: none;
    cursor: pointer;
    margin: 0 0.125rem;
}

.Nav a.active {
    color: orange;
}

@container (min-width: 375px) {
    .Nav a {
        font-size: 1em;
    }
}