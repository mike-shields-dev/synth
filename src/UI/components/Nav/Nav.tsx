import css from './Nav.module.css';

interface Props {
    focus: string;
    updateFocus: (id: string) => void;
}

function Nav({ focus, updateFocus }: Props) {
    function onClick(e: React.MouseEvent<HTMLAnchorElement>) {
        const id = e.currentTarget.dataset.focus_id;
        
        updateFocus(id || "");
    }

    const links = [
        { id: "oscillator", displayName: "Oscillator" },
        { id: "filter", displayName: "Filter" },
        { id: "filterEnvelope", displayName: "Filter Env" },
        { id: "ampEnvelope", displayName: "Amp Env" },
    ]

    return (
        <nav className={css.Nav}>
            {
                links.map(({ id, displayName }) =>
                    <a
                        className={css[focus === id ? "active" : ""]}
                        key={`nav-${id}-anchor`}
                        onClick={onClick}
                        data-focus_id={id}
                        href={`#${id}`}
                        aria-label={`Go to ${id} section`}
                    >
                        {displayName}
                    </a>
            )}
        </nav>
    )
}

export default Nav;
