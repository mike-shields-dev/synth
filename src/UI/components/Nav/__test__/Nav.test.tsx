import Nav from '../Nav';
import { render, screen, fireEvent } from '@testing-library/react';
import { vi } from 'vitest';
 
const validProps = {
    focus: 'oscillator',
    updateFocus: vi.fn()
}

describe('Nav', () => {
    it('matches snapshot', () => {
        const { container } = render(<Nav {...validProps} />);

        expect(container).toMatchSnapshot();
    });

    it('renders a link for each section in the Synth controls', () => {
        render(<Nav {...validProps} />)
        
        const links = [
            { id: 'oscillator', displayName: 'Oscillator' },
            { id: 'filter', displayName: 'Filter' },
            { id: 'filterEnvelope', displayName: 'Filter Env' },
            { id: 'ampEnvelope', displayName: 'Amp Env' },
        ];

        links.forEach(link =>
            expect(screen.getByRole('link', {
                name: `Go to ${link.id} section`
            })).toBeInTheDocument()
        );
    });

    it('each link causes the correct synth controls section to be focused when clicked', async () => {
        render(<Nav {...validProps} />)

        const links = [
            { id: 'oscillator', displayName: 'Oscillator' },
            { id: 'filter', displayName: 'Filter' },
            { id: 'filterEnvelope', displayName: 'Filter Env' },
            { id: 'ampEnvelope', displayName: 'Amp Env' },
        ];
        
        for (const link of links) {
            fireEvent.click(screen.getByRole('link', {
                name: `Go to ${link.id} section`
            }));

            expect(validProps.updateFocus).toHaveBeenCalledWith(link.id);
        } 
    });
});