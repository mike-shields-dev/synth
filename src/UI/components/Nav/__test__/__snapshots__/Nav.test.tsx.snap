// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Nav > matches snapshot 1`] = `
<div>
  <nav
    class="_Nav_03786e"
  >
    <a
      aria-label="Go to oscillator section"
      class="_active_03786e"
      data-focus_id="oscillator"
      href="#oscillator"
    >
      Oscillator
    </a>
    <a
      aria-label="Go to filter section"
      class="__03786e"
      data-focus_id="filter"
      href="#filter"
    >
      Filter
    </a>
    <a
      aria-label="Go to filterEnvelope section"
      class="__03786e"
      data-focus_id="filterEnvelope"
      href="#filterEnvelope"
    >
      Filter Env
    </a>
    <a
      aria-label="Go to ampEnvelope section"
      class="__03786e"
      data-focus_id="ampEnvelope"
      href="#ampEnvelope"
    >
      Amp Env
    </a>
  </nav>
</div>
`;
